clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,355 @@
+            newAssignment.addAll(diffPartitions);
+            return sortPartitions(newAssignment);
+        };
+
+        return reconcile(assignment,
+                String.format("%s.onPartitionsAssigned()", ConsumerRebalanceListener.class.getSimpleName()),
+                setDifferenceGenerator,
+                AssignPartitionsEvent::new,
+                modifier);
+    }
+
+    private ReconciliationResult reconcile(Optional<Assignment> assignment,
+                                           String listenerMethodName,
+                                           SetDifferenceGenerator setDifferenceGenerator,
+                                           EventGenerator eventGenerator,
+                                           AssignmentGenerator assignmentGenerator) {
+        // "diff" the two sets of partitions: our "current" assignment and the "target" assignment. The result is
+        // sorted primarily so when the partitions show up in the logs, it's easier for us humans to understand.
+        SortedSet<TopicPartition> diffPartitions = setDifferenceGenerator.generate(
+                subscriptions.assignedPartitions(),
+                targetPartitions(metadata, assignment)
+        );
+
+        if (diffPartitions.isEmpty()) {
+            log.debug("Skipping invocation of {} as no partitions were changed in the new assignment",
+                    listenerMethodName);
+            return ReconciliationResult.UNCHANGED;
+        }
+
+        // Set up our callback invocation. We don't block here waiting on it its completion, though.
+        log.debug("Preparing to invoke {} with the following partitions: {}", listenerMethodName, diffPartitions);
+        RebalanceCallbackEvent event = eventGenerator.generate(diffPartitions);
+        inflightCallback = Optional.of(event);
+
+        // Enqueue it in our background->application shared queue. This should be invoked in the Consumer.poll() method.
+        backgroundEventQueue.add(event);
+
+        // It is only after the appropriate ConsumerRebalanceListener has been executed on the application thread
+        // (via Consumer.poll()), can we "commit" the change to the assigned partitions in SubscriptionState.
+        event.future().whenComplete((result, error) -> {
+            if (error != null) {
+                log.warn("An error occurred when invoking {} with the following partitions: {}",
+                        listenerMethodName,
+                        diffPartitions,
+                        error);
+                // TODO: should we proceed or abort?
+            }
+
+            log.debug("{} was successfully executed with the following partitions: {}",
+                    listenerMethodName,
+                    diffPartitions);
+
+            SortedSet<TopicPartition> newAssignment = assignmentGenerator.generate(diffPartitions);
+
+            // And finally... assign the new set of partitions!
+            subscriptions.assignFromSubscribed(newAssignment);
```

Comment:
Just for the record, this is the link between this component and the fetching logic, based on the existing subscription state. I expect that after this line, the next iteration of the poll loop will start fetching from the new partitions (if not paused and if they have a valid position)

Reply:
Correct.